src/internal.c

```diff
@@ -5055,19 +5055,46 @@ int EccSharedSecret(WOLFSSL* ssl, ecc_key* priv_key, ecc_key* pub_key,
         }
     }
 
-    /* Handle async pending response */
+#ifdef WOLFSSL_ECDHE_NONBLOCK
+    if (ret == FP_WOULDBLOCK) {
+        ret = WC_PENDING_E;
+    }
+#endif /* WOLFSSL_ECDHE_NONBLOCK */
+
 #ifdef WOLFSSL_ASYNC_CRYPT
-    if (ret == WC_PENDING_E) {
-        ret = wolfSSL_AsyncPush(ssl, asyncDev);
+    /* Handle async pending response */
+
+    /*
+     * Don't add a new async event if there's already a non-blocking operation
+     * in progress.
+     */
+    if (!nonblockInProg) {
+        if (ret == WC_PENDING_E) {
+            ret = wolfSSL_AsyncPush(ssl, &priv_key->asyncDev);
+        }
+    }
+    else {
+    #ifdef WOLFSSL_ECDHE_NONBLOCK
+        if (ret == 0) {
+            ret = wolfEventQueue_Remove(&ssl->ctx->event_queue, &ssl->asyncDev->event);
+            if (ret == 0) {
+                ret = wolfSSL_AsyncPop(ssl, &ssl->options.asyncState);
+            }
+        }
+    #else
+        ret = NOT_COMPILED_IN;
+    #endif /* WOLFSSL_ECDHE_NONBLOCK */
     }
+#else
+    (void)nonblockInProg;
 #endif /* WOLFSSL_ASYNC_CRYPT */
 
     WOLFSSL_LEAVE("EccSharedSecret", ret);
 
     return ret;
 }
 
-int EccMakeKey(WOLFSSL* ssl, ecc_key* key, ecc_key* peer)
+int EccMakeKey(WOLFSSL* ssl, ecc_key* key, ecc_key* peer, byte nonblockInProg)
```

Comment:
This is only ECC make key and shared secret, but it will also be needed in sign and verify.